### PR TITLE
Feature/minimal api

### PR DESCRIPTION
**Pull request by ktfth**
As described on the issue #518 the minimal api was made based on the single responsabilitie principle to work with a minimal version of the copy behavior without an need of cut because is unecessary.

### CODE REVIEW

File: demo/minimal-api.html [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+<html lang="en" dir="ltr">
+  <head>
+    <meta charset="utf-8">
+    <title></title>
+  </head>
+  <body>
+    <button id="btn" type="button">Copy</button>
+    <script src="../dist-minimal/clipboard.min.js"></script>
+    <script>
+      let btn = document.getElementById('btn');
```
Comment (helderburato) [inline demo/minimal-api.html:11]: It doesn't will cause a great impact, but you can use a `const` to create the BTN instance.

File: minimal-webpack.config.js [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    entry: path.resolve(__dirname, 'src') + '/clipboard-minimal.js',
+    mode: 'production',
+    output: {
+        filename: production ? 'clipboard.min.js' : 'clipboard.js',
+        path: path.resolve(__dirname, './dist-minimal'),
+        library: 'clipboard',
+        globalObject: 'this',
+        libraryExport: 'default',
+        // libraryTarget: 'umd',
+        libraryTarget: 'window',
```
Comment (helderburato) [inline minimal-webpack.config.js:23]: What do you think about to remove this unused libraryTarget from code and maintain just the `window`?

File: src/clipboard-minimal.js [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+      fakeElem.style.margin = '0';
+      // Move element out of screen horizontally
+      fakeElem.style.position = 'absolute';
+      fakeElem.style[ isRTL ? 'right' : 'left' ] = '-9999px';
+      // Move element to the same position vertically
+      let yPosition = window.pageYOffset || document.documentElement.scrollTop;
+      fakeElem.style.top = `${yPosition}px`;
+
+      fakeElem.setAttribute('readonly', '');
+      fakeElem.value = input;
```
Comment (helderburato) [inline src/clipboard-minimal.js:21]: Maybe this part could be separated to a single method, something like `generateFakeElement`, what do you think about?

**Review by helderburato — CHANGES_REQUESTED** [review]
I gave some suggestions I just requesting changes because Travis throw some error.

File: src/clipboard-minimal.js (hunk 3 above)
Comment (vitormalencar) [reply src/clipboard-minimal.js:21]: also, another point here It does like duplication logic here, 	we have the same logic in the https://github.com/zenorocha/clipboard.js/blob/master/src/clipboard-action.js#L47 

I do believe that we can avoid that

File: package.json [hunk 4]
```diff
@@ -45,6 +45,7 @@
     "build-debug": "webpack",
     "build-min": "cross-env NODE_ENV=production webpack",
     "build-watch": "webpack --watch",
+    "build-minimal-api": "cross-env NODE_ENV=production webpack --config ./minimal-webpack.config.js",
```
Comment (vitormalencar) [inline package.json:48]: another small nit here is that we should have one build command instead of 2

File: .gitignore [hunk 5]
```diff
@@ -4,3 +4,4 @@ bower_components
 node_modules
 yarn-error.log
 yarn.lock
+package-lock.json
```
Comment (vitormalencar) [inline .gitignore:7]: `package-lock.json` is intended to be checked into source control. You should commit this file. According to npm help package-lock.json

https://github.com/npm/cli/blob/release-6.14.7/docs/content/configuring-npm/package-lock-json.md

**Review by vitormalencar — CHANGES_REQUESTED** [review]
Nice job @ktfth here, 
I like the approach of creating a minimal API for interaction, however, we need to create this inside of the project and not as an isolated file, what I think is nice is that we can use this PR to refactor the core lib

File: src/clipboard-minimal.js (hunk 3 above)
Comment (ktfth) [reply src/clipboard-minimal.js:21]: Have not the same behavior of the clipboard action on that case, we follow the same principle of other implementations used as reference.

File: minimal-webpack.config.js (hunk 2 above)
Comment (ktfth) [reply minimal-webpack.config.js:23]: I can can change that

File: demo/minimal-api.html (hunk 1 above)
Comment (ktfth) [reply demo/minimal-api.html:11]: Deal, the change can be made.

File: package.json (hunk 4 above)
Comment (ktfth) [reply package.json:48]: The webpack do not build to source of files separated, the minimal API have another approach here.

File: .gitignore (hunk 5 above)
Comment (ktfth) [reply .gitignore:7]: Ok, the file was added on the way

File: minimal-webpack.config.js [hunk 6]
```diff
@@ -0,0 +1,45 @@
+const pkg = require('./package.json');
+const path = require('path');
+const webpack = require('webpack');
+const UglifyJSPlugin = require('uglifyjs-webpack-plugin');
+
+const production = process.env.NODE_ENV === 'production' || false;
```
Comment (obetomuniz) [inline minimal-webpack.config.js:6]: Only `process.env.NODE_ENV === 'production'` will work here.

File: src/clipboard-minimal.js [hunk 7]
```diff
@@ -0,0 +1,31 @@
+function clipboard() {
+  return {
+    copy: (input) => {
+      const isRTL = document.documentElement.getAttribute('dir') == 'rtl';
```
Comment (obetomuniz) [inline src/clipboard-minimal.js:4]: What about use strict equality here?

File: src/clipboard-minimal.js (hunk 3 above)
Comment (obetomuniz) [reply src/clipboard-minimal.js:21]: I agreed that we can at least wrap the code duplicated in a short helper. Just to improve maintainability. WDYT @ktfth @vitormalencar ?

**Comment by obetomuniz** [comment]
Great job here @ktfth, I just added a few nits, but consider the requests from @vitormalencar and @helderburato . Also, I'd agree with @vitormalencar about move it into the project, which means that we need a bit more discussions related to the best way to increment the project with more code without attack the current issues about bundle size and outdated approaches using classes. I would suggest putting this PR on hold until we solve such things and revisit it right after such work. What do you guys think?

File: package.json (hunk 4 above)
Comment (obetomuniz) [reply package.json]: I'd agree with @vitormalencar, so I would not suggest this approach here @ktfth, it goes against having a single source of truth inside the project. The idea to have a tiny API to operate the library programmatically is fabulous, but it should live inside the main build. Anyway, thinking about code-splitting in our future architecture, export only a `copy()`/`cut()` method will be awesome in terms of performance.

**Comment by vitormalencar** [comment]
Closing for now we can use this as a future reference =P

**Pull request closed without merging**